xds/internal/balancer/gracefulswitch/gracefulswitchbalancer.go

```diff
@@ -0,0 +1,398 @@
+	bOpts balancer.BuildOptions
+	cc    balancer.ClientConn
+
+	mu               sync.Mutex
+	balancerCurrent  balancer.Balancer
+	balancerPending  balancer.Balancer
+	scToSubBalancer  map[balancer.SubConn]balancer.Balancer
+	pendingState     balancer.State
+	currentLbIsReady bool
+
+	// currentMu protects operations on the current balancer.
+	currentMu sync.Mutex
+
+	closed *grpcsync.Event
+}
+
+func (gsb *gracefulSwitchBalancer) updateState(bal balancer.Balancer, state balancer.State) {
+	if gsb.closed.HasFired() {
+		return
+	}
+
+	// I want this to not intersplice with other updateState() calls and for
+	// this to happen atomically (i.e. if this update isn't atomic writing to
+	// pending state concurrently and sending wrong one out for one
+	// updateState() call).
+	gsb.mu.Lock()
+	defer gsb.mu.Unlock()
+
+	if !gsb.balancerCurrentOrPending(bal) {
+		return
+	}
+
+	if bal == gsb.balancerPending {
+		// Cache the pending state and picker if you don't need to send it at
+		// this instant (i.e. the LB policy is connecting). This balancer can
+		// send it later on the current LB exiting READY.
+		gsb.pendingState = state
+		// "If the channel is currently in a state other than READY, the new
+		// policy will be swapped into place immediately." - Java
+		// implementation.
+		if state.ConnectivityState != connectivity.Connecting || !gsb.currentLbIsReady {
+			gsb.swap()
+		}
+		return
+	}
+	gsb.currentLbIsReady = state.ConnectivityState == connectivity.Ready
+	// In the specific case that the current lb exits ready, and there is a
+	// pending lb, you can forward the pending LB's cached State up to
+	// ClientConn and swap the pending into the current. "Otherwise, the
+	// channel will keep using the old policy until...the old policy exits
+	// READY." - Java
+	if !gsb.currentLbIsReady && gsb.balancerPending != nil {
+		gsb.swap()
+		return
+	}
+	// Even if there is a pending balancer waiting to be gracefully
+	// switched to, still forward current balancer updates to the Client
+	// Conn. Ignoring state + picker from the current would cause
+	// undefined behavior/cause the system to behave incorrectly from
+	// the current LB policies perspective. Also, the current LB is
+	// still being used by grpc to choose SubConns per RPC, and thus
+	// should use the most updated form of the current balancer.
+	gsb.cc.UpdateState(state)
+}
+
+// swap swaps out the current lb with the pending LB and updates the ClientConn.
+// The caller must hold gsb.mu.
+func (gsb *gracefulSwitchBalancer) swap() {
+	gsb.cc.UpdateState(gsb.pendingState)
+	currBalToClose := gsb.balancerCurrent
+	go func() {
+		gsb.currentMu.Lock()
+		defer gsb.currentMu.Unlock()
+		currBalToClose.Close()
+		gsb.mu.Lock()
+		defer gsb.mu.Unlock()
+		for sc, bal := range gsb.scToSubBalancer {
+			if bal == gsb.balancerCurrent {
+				gsb.cc.RemoveSubConn(sc)
+				delete(gsb.scToSubBalancer, sc)
+			}
+		}
+
+		gsb.balancerCurrent = gsb.balancerPending
+		gsb.balancerPending = nil
```

Comment:
It's important that these operations happen before `swap()` returns.  (Imagine if this goroutine gets blocked for hours and lots of other stuff happens, including several new balancers being swapped between.)